test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,1167 @@
+
+    /* initialize buffer with requested params */
+    par->type_offset = lb;      /* convert offset to bytes */
+    err = MPI_Type_get_name(basic_type, basic_type_name, &len);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    DTPI_Type_init_buf(par, basic_type_name, buf);
+
+    dtp->DTP_obj_array[obj_idx].DTP_obj_buf = buf;
+    dtp->DTP_obj_array[obj_idx].private_info = dtpi;
+
+    /* initialize private datatype info data */
+    dtpi->obj_type = DTPI_OBJ_TYPE__INDEXED;
+    dtpi->type_basic_size = basic_type_size;
+    dtpi->type_extent = extent;
+    dtpi->type_lb = lb;
+    dtpi->type_ub = lb + extent * type_count;
+    dtpi->u.indexed.displs = type_displs;       /* freed in DTP_obj_free() */
+    dtpi->u.indexed.num_blocks = type_blklens;  /* freed in DTP_obj_free() */
+
+    /* set type name for debug information */
+    sprintf(type_name, "%s (%s basic_type %li count %li blocklen %li stride)", "indexed",
+            basic_type_name, type_count, type_blklen, type_stride);
+    err = MPI_Type_set_name(dtp->DTP_obj_array[obj_idx].DTP_obj_type, (char *) type_name);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+
+  fn_exit:
+    return err;
+
+  fn_fail:
+    if (dtpi) {
+        free(dtpi);
+    }
+    if (type_displs) {
+        free(type_displs);
+    }
+    if (type_blklens) {
+        free(type_blklens);
+    }
+    goto fn_exit;
+}
+
+int DTPI_Block_indexed_create(struct params *par, DTP_t dtp)
+{
+    int i, err = DTP_SUCCESS;
+    int len, obj_idx;
+    int *type_displs = NULL;
+    int basic_type_size;
+    MPI_Aint type_offset, type_count, type_stride, type_blklen;
+    MPI_Aint size;
+    MPI_Aint extent, lb;
+    MPI_Datatype basic_type;
+    void *buf;
+    DTPI_t *dtpi = NULL;
+    char type_name[TYPE_NAME_MAXLEN] = { 0 };
+    char basic_type_name[BASIC_TYPE_NAME_MAXLEN] = { 0 };
+
+    /* allocate space for private datatype info */
+    dtpi = (DTPI_t *) malloc(sizeof(*dtpi));
+    if (!dtpi) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    type_count = par->type_count;
+    type_displs = (int *) malloc(sizeof(*type_displs) * type_count);
+    if (!type_displs) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    type_offset = par->type_offset;
+    type_stride = par->type_stride;
+    for (i = 0; i < type_count; i++) {
+        type_displs[i] = type_offset + type_stride * i;
+    }
+
+    /* allocate buffer */
+    basic_type = dtp->DTP_type_signature.DTP_pool_basic.DTP_basic_type;
+    err = MPI_Type_size(basic_type, &basic_type_size);
+    if (err) {
+        DTPI_Print_error(err);
+    }
+    lb = type_offset * basic_type_size;
+    extent = type_stride * basic_type_size;
+    size = lb + extent * type_count;
+    buf = (void *) malloc(size);
```

Comment:
Same here, please use MPI_Type_get_extent instead of manually calculating buffer size.